File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let minPrice = prices[0]
    let maxProfit = 0
    for (let i = 0; i < prices.length; i++) {
        if (prices[i] < minPrice) {
            minPrice = prices[i]
        }
        
        let potentialProfit = prices[i] - minPrice
        
        if (potentialProfit > maxProfit) {
            maxProfit = potentialProfit
        }
    }
    
    return maxProfit
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minPrice = prices[0]
        maxProfit = 0
        
        for i in range(1, len(prices)):
            if prices[i] < minPrice:
                minPrice = prices[i]
            
            potentialProfit = prices[i] - minPrice
            
            if potentialProfit > maxProfit:
                maxProfit = potentialProfit
        
        return maxProfit